runner/cf-driver/prepare.sh

```diff
@@ -38,6 +38,76 @@ start_container () {
        --vars-file "$TMPVARFILE"
 }
 
+start_service () {
+    alias_name="$1"
+    container_id="$2"
+    image_name="$3"
+    container_entrypoint="$4"
+    container_command="$5"
+
+    if [ -z "$container_id" ] || [ -z "$image_name" ]; then
+       echo 'Usage: start_service CONTAINER_ID IMAGE_NAME CONTAINER_ENTRYPOINT CONTAINER_COMMAND'
+       exit 1
+    fi
+    if [ -n "$container_entrypoint" ] || [ -n "$container_command" ]; then
+        # TODO - cf push allows use of -c or --start-command but not a separate
+        # entrypoint. May need to add logic to gracefully convert entrypoint to
+        # a command.
+        echo 'container_entrypoint and container_command are not yet supported in services - Sorry!'
+        exit 1
+    fi
+
+    if cf app --guid "$container_id" >/dev/null 2>/dev/null ; then
+        echo 'Found old instance of runner service, deleting'
+        cf delete "$container_id"
+    fi
+
+    # TODO - Figure out how to handle command and non-global memory definition
+    cf push "$container_id" --docker-image "$image_name" -m "$WORKER_MEMORY" \
+        --no-route --health-check-type process
+
+    echo "DEBUG: cf map-route $container_id apps.internal --hostname $alias_name"
+    cf map-route "$container_id" apps.internal --hostname "$container_id"
+}
+
+allow_access_to_service () {
+    source_app="$1"
+    destination_service_app="$2"
+    current_org=$(echo "$VCAP_APPLICATION" | jq --raw-output ".organization_name")
+    current_space=$(echo "$VCAP_APPLICATION" | jq --raw-output ".space_name")
+
+    # TODO NOTE: This is foolish and allows all TCP ports for now.
+    # This is limiting and sloppy.
+    protocol="tcp"
+    ports="20-10000"
+
+    cf add-network-policy "$source_app" \
+        --destination-app "$destination_service_app" \
+        -o "$current_org" -s "$current_space" \
+        --protocol "$protocol" --port "$ports"
+}
+
+start_services () {
+    container_id_base="$1"
+    ci_job_services="$2"
+
+    if [ -z "$ci_job_services" ]; then
+       echo "No services defined in ci_job_services - Skipping service startup"
+       return
+    fi
+
+    for l in $(echo "$ci_job_services" | jq -rc '.[]'); do
+        alias_name=$(echo "$l" | jq -r '. | .alias | select(. != null)')
+        container_id="${container_id_base}-svc-${alias_name}"
+        image_name=$(echo "$l" | jq -r '. | .name | select(. != null)')
+        container_entrypoint=$(echo "$l" | jq -r '. | .entrypoint| select(. != null)')
+        container_command=$(echo "$l" | jq -r '. | .command | select(. != null)')
```

Comment:
I don't like this but since I need to avoid any of the values being converted from `nuil` to `"null"` this is the ugly I settled on.

Reply:
I'm wondering if you can drop the first pipe, and, unless you want `false` values, you could simplify `select`, like:

```bash
image_name=$(echo "$l" | jq -r '.name | select(.)')
```

This is what I mean: https://jqplay.org/s/rR4K2KodIO0F2w_V

It _will_ give you a refence error if you try to access a bool, number or array though. So if that's a possibility and you don't want it to error out you could add a `?` at the end like: `.name? | select(.)`.

Reply:
I like it! I used the `?` form for the two items that are optional and left it off `.name` and `.alias` to blow things up if they are missing.

Reply:
I'm not sure it will blow up if they're missing--a null or undefined shouldn't result in an error, but I'm also not sure if that's important.

Reply:
Bah, it won't. I should have tested. I guess I lied!

New push uses `-er` on the alias and name to require those. (Missing, false, or null should exit 1 resulting in bash to blow up since we are using `set -e` in the script.)

I added some commentary to explain.